File: perl/2015/day04b/day04b.pl
#!/usr/bin/env perl

use strict;
use warnings;

use Digest::MD5 qw(md5_hex);

sub usage {
    my $progname = $0;
    print "usage: $progname <key>\n";
    exit 1;
}

sub process {
    my ($key) = @_;
    my $n = 1;
    while ( 1 ) {
        my $tryKey = $key . $n;
        my $digest = md5_hex($tryKey);
        if ( $digest =~ /^000000/ ) {
            return $n;
        } else {
            $n++;
        }
    }
}

sub main {
    if ( @ARGV < 1 ) {
        usage();
    }
    my $key = $ARGV[0];
    my $result = process($key);
    print "result = $result\n";
}

main();

File: r/2015/day04b/day04b.r
#!/usr/bin/env Rscript

library(digest)

usage <- function() {
    args <- commandArgs()
    progname <- sub("^--file=", "", args[grep("^--file=", args)])
    cat("usage:", progname, "<input file>\n")
    quit(status = 1)
}

process <- function(key) {
    n <- 0
    repeat {
        tryKey <- paste0(key, n)
        hexDigest <- digest(tryKey, algo = "md5", serialize = FALSE)
        if ( startsWith(hexDigest, "000000") ) {
            return(n)
        }
        n <- n + 1
    }
}

main <- function() {
    args <- commandArgs(trailingOnly = TRUE)
    if ( length(args) < 1 ) {
        usage()
    }
    key <- args[1]
    result <- process(key)
    cat("result =", result, "\n")
}

main()
